File: runinator-command-center/src/core/utils/key-value-object.ts
import type { JsonRecord } from "../domain/models";
import { asJsonValue } from "../domain/json";

export interface RenameObjectKeyResult {
  value: JsonRecord;
  error: string;
}

export function uniqueObjectKey(record: JsonRecord, base = "key"): string {
  let index = 1;
  let key = base;

  while (Object.prototype.hasOwnProperty.call(record, key)) {
    index += 1;
    key = `${base}_${String(index)}`;
  }

  return key;
}

export function renameObjectKey(
  record: JsonRecord,
  previousKey: string,
  nextKey: string,
): RenameObjectKeyResult {
  const trimmed = nextKey.trim();

  if (!trimmed) {
    return { value: record, error: "Key is required" };
  }

  if (trimmed !== previousKey && Object.prototype.hasOwnProperty.call(record, trimmed)) {
    return { value: record, error: "Key already exists" };
  }

  if (trimmed === previousKey) {
    return { value: record, error: "" };
  }

  const next: JsonRecord = {};

  for (const [key, value] of Object.entries(record)) {
    next[key === previousKey ? trimmed : key] = value;
  }

  return { value: next, error: "" };
}

export function setObjectValue(record: JsonRecord, key: string, value: unknown): JsonRecord {
  if (!key.trim()) {
    return record;
  }

  return { ...record, [key]: asJsonValue(value) };
}

export function removeObjectKey(record: JsonRecord, key: string): JsonRecord {
  return Object.fromEntries(Object.entries(record).filter(([entryKey]) => entryKey !== key));
}

File: runinator-command-center-flutter/lib/core/utils/key_value_object.dart
// port of core/utils/key-value-object.ts.

import '../domain/json.dart';

class RenameObjectKeyResult {
  const RenameObjectKeyResult({required this.value, required this.error});

  final JsonRecord value;
  final String error;
}

String uniqueObjectKey(JsonRecord record, [String base = 'key']) {
  var index = 1;
  var key = base;

  while (record.containsKey(key)) {
    index += 1;
    key = '${base}_$index';
  }

  return key;
}

RenameObjectKeyResult renameObjectKey(JsonRecord record, String previousKey, String nextKey) {
  final trimmed = nextKey.trim();

  if (trimmed.isEmpty) {
    return RenameObjectKeyResult(value: record, error: 'Key is required');
  }

  if (trimmed != previousKey && record.containsKey(trimmed)) {
    return RenameObjectKeyResult(value: record, error: 'Key already exists');
  }

  if (trimmed == previousKey) {
    return RenameObjectKeyResult(value: record, error: '');
  }

  final next = <String, Object?>{
    for (final entry in record.entries) (entry.key == previousKey ? trimmed : entry.key): entry.value,
  };

  return RenameObjectKeyResult(value: next, error: '');
}

JsonRecord setObjectValue(JsonRecord record, String key, Object? value) {
  if (key.trim().isEmpty) {
    return record;
  }

  return {...record, key: asJsonValue(value)};
}

JsonRecord removeObjectKey(JsonRecord record, String key) => {
      for (final entry in record.entries)
        if (entry.key != key) entry.key: entry.value,
    };
